File: sdk/node/src/worker.ts
import type {
  Msg,
  NatsConnection,
  Subscription,
  SubscriptionOptions,
} from "nats";
import type { Type } from "protobufjs";
import { loadRoot, SUBJECT_RESULT, DEFAULT_PROTOCOL_VERSION } from "./protos";
import { encodeDeterministic, encodeUnsignedForSignature } from "./codec";
import type { Logger } from "./logger";
import type { MetricsHook } from "./metrics";
import { noopMetrics } from "./metrics";
import * as crypto from "crypto";
import { verifyInboundPacket } from "./security";

/** Callback that processes a decoded JobRequest and returns a JobResult. */
type Handler = (jobRequest: any) => Promise<any>;

/** Middleware wraps a worker Handler, returning a new Handler. */
export type WorkerMiddleware = (next: Handler) => Handler;

/** Configuration for {@link startWorker}. */
export interface WorkerConfig {
  nc: NatsConnection;
  subject: string;
  queue?: string;
  handler: Handler;
  publicKeyMap?: { [senderId: string]: string }; // senderId -> public key in PEM format
  privateKey?: string; // private key in PEM format for signing outgoing messages
  senderId: string;
  /** Optional middleware applied in FIFO order before the handler. */
  middlewares?: WorkerMiddleware[];
  logger?: Logger;
  metrics?: MetricsHook;
}

interface JobRequestView {
  jobId?: string;
  topic?: string;
}

interface PacketView {
  traceId?: string;
  jobRequest?: JobRequestView;
}

interface SignablePacket {
  signature?: Uint8Array;
}

interface WorkerTypes {
  busPacket: Type;
  jobResult: Type;
}

/**
 * Subscribes to a NATS subject and dispatches incoming JobRequests to the handler.
 *
 * Results are wrapped in a signed BusPacket and published to the result subject.
 *
 * @param cfg - Worker configuration including NATS connection and handler.
 * @returns The underlying NATS subscription.
 */
export async function startWorker(cfg: WorkerConfig): Promise<Subscription> {
  const root = await loadRoot();
  const types: WorkerTypes = {
    busPacket: root.lookupType("cordum.agent.v1.BusPacket"),
    jobResult: root.lookupType("cordum.agent.v1.JobResult"),
  };
  const logger: Logger = cfg.logger ?? console;
  const metrics: MetricsHook = cfg.metrics ?? noopMetrics;
  const callback: NonNullable<SubscriptionOptions["callback"]> = (error, msg) => {
    if (error) {
      logger.error("worker subscription error", { error: String(error) });
      return;
    }
    void processWorkerMessage(cfg, types, logger, metrics, msg).catch((processingError) => {
      logger.error("worker message processing failed", {
        error: String(processingError),
      });
    });
  };
  return cfg.nc.subscribe(cfg.subject, {
    queue: cfg.queue ?? cfg.subject,
    callback,
  });
}

async function processWorkerMessage(
  cfg: WorkerConfig,
  types: WorkerTypes,
  logger: Logger,
  metrics: MetricsHook,
  msg: Msg
): Promise<void> {
  const packet = types.busPacket.decode(msg.data) as unknown as PacketView;
  try {
    verifyInboundPacket(types.busPacket, packet, cfg.publicKeyMap);
  } catch (error) {
    logger.warn("worker rejected inbound packet", { error: String(error) });
    return;
  }
  const request = packet.jobRequest;
  if (!request) return;
  metrics.onJobReceived(request.jobId ?? "", request.topic ?? "");
  const startedAt = Date.now();
  const result = await executeHandler(cfg, request);
  const elapsedMs = Date.now() - startedAt;
  recordResultMetrics(metrics, result, elapsedMs);
  await publishWorkerResult(cfg, types, packet, result);
}

async function executeHandler(
  cfg: WorkerConfig,
  request: JobRequestView
): Promise<Record<string, unknown>> {
  let wrappedHandler = cfg.handler;
  const middlewares = cfg.middlewares ?? [];
  for (let index = middlewares.length - 1; index >= 0; index -= 1) {
    wrappedHandler = middlewares[index](wrappedHandler);
  }
  let value: unknown;
  try {
    value = await wrappedHandler(request);
  } catch (error) {
    value = failureResult(request.jobId, error instanceof Error ? error.message : String(error));
  }
  const result = isRecord(value)
    ? value
    : failureResult(request.jobId, "handler returned null");
  if (typeof result.jobId !== "string" || result.jobId.length === 0) {
    result.jobId = request.jobId ?? "";
  }
  if (typeof result.workerId !== "string" || result.workerId.length === 0) {
    result.workerId = cfg.senderId;
  }
  return result;
}

function failureResult(jobId: string | undefined, errorMessage: string): Record<string, unknown> {
  return { jobId: jobId ?? "", status: "JOB_STATUS_FAILED", errorMessage };
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function recordResultMetrics(
  metrics: MetricsHook,
  result: Record<string, unknown>,
  elapsedMs: number
): void {
  const jobId = String(result.jobId ?? "");
  if (result.status === "JOB_STATUS_FAILED") {
    metrics.onJobFailed(jobId, String(result.errorMessage ?? ""));
    return;
  }
  metrics.onJobCompleted(
    jobId,
    elapsedMs,
    String(result.status ?? "JOB_STATUS_SUCCEEDED")
  );
}

async function publishWorkerResult(
  cfg: WorkerConfig,
  types: WorkerTypes,
  packet: PacketView,
  result: Record<string, unknown>
): Promise<void> {
  const resultMessage = types.jobResult.fromObject(result);
  const outgoing = types.busPacket.fromObject({
    traceId: packet.traceId,
    senderId: cfg.senderId,
    protocolVersion: DEFAULT_PROTOCOL_VERSION,
    createdAt: { seconds: Math.floor(Date.now() / 1000), nanos: 0 },
    jobResult: resultMessage,
  }) as unknown as SignablePacket;
  if (cfg.privateKey) {
    const signer = crypto.createSign("sha256");
    signer.update(encodeUnsignedForSignature(types.busPacket, outgoing));
    outgoing.signature = signer.sign(cfg.privateKey);
  }
  const data = encodeDeterministic(types.busPacket, outgoing);
  await cfg.nc.publish(SUBJECT_RESULT, data);
}

File: sdk/rust/src/worker.rs
use std::collections::HashMap;
use std::sync::Arc;
use std::time::{Instant, SystemTime};

use p256::ecdsa::{SigningKey, VerifyingKey};
use prost::Message;

use crate::codec::marshal_deterministic;
use crate::errors::CapError;
use crate::metrics::{MetricsHook, NoopMetrics};
use crate::middleware::{apply_middleware, Context, HandlerFn, MiddlewareFn};
use crate::pb::{bus_packet::Payload, BusPacket, JobRequest, JobResult, JobStatus};
use crate::signing::{sign_packet, verify_packet_signature};
use crate::subjects;

pub struct Worker {
    nc: async_nats::Client,
    subject: String,
    sender_id: String,
    handler: HandlerFn,
    middlewares: Vec<MiddlewareFn>,
    private_key: Option<SigningKey>,
    public_keys: HashMap<String, VerifyingKey>,
    metrics: Arc<dyn MetricsHook>,
}

impl Worker {
    pub fn new(
        nc: async_nats::Client,
        subject: String,
        sender_id: String,
        handler: HandlerFn,
    ) -> Self {
        Worker {
            nc,
            subject,
            sender_id,
            handler,
            middlewares: Vec::new(),
            private_key: None,
            public_keys: HashMap::new(),
            metrics: Arc::new(NoopMetrics),
        }
    }

    pub fn with_private_key(mut self, key: SigningKey) -> Self {
        self.private_key = Some(key);
        self
    }

    pub fn with_public_keys(mut self, keys: HashMap<String, VerifyingKey>) -> Self {
        self.public_keys = keys;
        self
    }

    pub fn with_metrics(mut self, metrics: Arc<dyn MetricsHook>) -> Self {
        self.metrics = metrics;
        self
    }

    pub fn use_middleware(&mut self, mw: MiddlewareFn) {
        self.middlewares.push(mw);
    }

    pub async fn start(&self) -> Result<(), CapError> {
        let wrapped = apply_middleware(self.handler.clone(), &self.middlewares);

        let mut sub = self
            .nc
            .queue_subscribe(self.subject.clone().into(), self.subject.clone().into())
            .await
            .map_err(|e| CapError::subscribe_failed(&e.to_string()))?;

        let nc = self.nc.clone();
        let sender_id = self.sender_id.clone();
        let private_key = self.private_key.clone();
        let public_keys = self.public_keys.clone();
        let metrics = self.metrics.clone();

        tokio::spawn(async move {
            while let Some(msg) = sub.next().await {
                let packet = match BusPacket::decode(msg.payload.as_ref()) {
                    Ok(p) => p,
                    Err(e) => {
                        metrics.on_error("parse", &format!("failed to parse: {}", e));
                        continue;
                    }
                };

                if !public_keys.is_empty() {
                    if let Some(key) = public_keys.get(&packet.sender_id) {
                        if packet.signature.is_empty() {
                            continue;
                        }
                        if verify_packet_signature(&packet, key).is_err() {
                            continue;
                        }
                    } else {
                        continue;
                    }
                }

                let req = match &packet.payload {
                    Some(Payload::JobRequest(r)) => r.clone(),
                    _ => continue,
                };

                metrics.on_job_received(&req.job_id, &req.topic);

                let ctx = Context {
                    trace_id: packet.trace_id.clone(),
                    sender_id: packet.sender_id.clone(),
                    packet: Some(packet.clone()),
                };

                let start = Instant::now();
                let result = match wrapped(&ctx, &req) {
                    Ok(mut r) => {
                        if r.job_id.is_empty() {
                            r.job_id = req.job_id.clone();
                        }
                        if r.worker_id.is_empty() {
                            r.worker_id = sender_id.clone();
                        }
                        let elapsed = start.elapsed().as_millis() as u64;
                        if r.status == JobStatus::Failed as i32 {
                            // FAILED
                            metrics.on_job_failed(&r.job_id, &r.error_message);
                        } else {
                            metrics.on_job_completed(&r.job_id, elapsed, "SUCCEEDED");
                        }
                        r
                    }
                    Err(e) => {
                        metrics.on_job_failed(&req.job_id, &e.to_string());
                        JobResult {
                            job_id: req.job_id.clone(),
                            worker_id: sender_id.clone(),
                            status: JobStatus::Failed as i32,
                            error_message: e.to_string(),
                            ..Default::default()
                        }
                    }
                };

                let now = SystemTime::now()
                    .duration_since(SystemTime::UNIX_EPOCH)
                    .unwrap_or_default();

                let mut out = BusPacket {
                    trace_id: packet.trace_id.clone(),
                    sender_id: sender_id.clone(),
                    protocol_version: subjects::DEFAULT_PROTOCOL_VERSION,
                    created_at: Some(prost_types::Timestamp {
                        seconds: now.as_secs() as i64,
                        nanos: now.subsec_nanos() as i32,
                    }),
                    payload: Some(Payload::JobResult(result)),
                    ..Default::default()
                };

                if let Some(ref key) = private_key {
                    if sign_packet(&mut out, key).is_err() {
                        metrics.on_error("signing", "failed to sign outgoing packet");
                        continue;
                    }
                }

                let data = marshal_deterministic(&out);
                let _ = nc
                    .publish(subjects::SUBJECT_RESULT.into(), data.into())
                    .await;
            }
        });

        Ok(())
    }
}

use futures::StreamExt;
